### DISCUSSION TITLE
Surface Meshing Failing on Imported Exodus Meshes

### DISCUSSION THREAD

**Discussion by wrkendrick** [svalinn/DAGMC#956] [Q&A]
Hey all,

I'm attempting to take some pre-generated ExodusII meshes and convert them into DAGMC .h5m files in order to run them in OpenMC. I've run into a strange problem where I'm unable to perform surface Trimeshing on an exodus mesh-based geometry that I import. Cubit just segfaults when I try. This obviously stops me from having a facet geometry to export as .h5m!

The test geometry I'm using is a simple cylinder that I meshed, exported, then imported.

![image](https://github.com/svalinn/DAGMC/assets/43858737/0c335ff4-8897-4139-b38d-a72610f58bce)

Steps to recreate this problem:
- Import mesh geometry
- Delete existing mesh
- Apply trimesh scheme (set trimesher coarse on ratio 100 angle 5 (I've tried a variety of angles)) to all surfaces
- Attempt to mesh
- Segfault!

I'm running this on Coreform Cubit 2024.3. I also tried meshing a single surface at a time to ensure that it wasn't my RAM running out. I've tried this on a couple different exodus meshes to confirm. I've attached both the exodus mesh and the cubit geometry (that can be successfully meshed) so you can try to replicate this issue. 

Arguably this may be a Cubit issue, not DAGMC. However, generating DAGMC facet files from Exodus meshes feels common enough that I'm hoping someone here has a workaround!

Thanks y'all.

[geometry.zip](https://github.com/user-attachments/files/15887072/geometry.zip)

**Comment by pshriwise** [comment]
Hi @wrkendrick, I've reproduced this behavior. It appears it only occurs when using the coarse mesh settings, and, as you said, I think it has something to do with the meshing internals as opposed to the DAGMC exporter -- I was able to load an exodus mesh and export it to a DAGMC file (both with the existing mesh and re-meshsed surfaces without the coarse mesh settings). I'm going to

**Reply by pshriwise** [reply]
I'm going to post a link to this issue in Coreform's issue tracker [here](https://forum.coreform.com/c/cubit/9)

**Reply by wrkendrick** [reply]
Thanks Patrick! When you say you successfully exported the DAGMC file with the existing mesh and a remeshed-surface version:
-  By "with the existing mesh" do you mean you simply imported the exodus mesh, and then exported it to .h5m? I tried this once before but when I ran the geometry as my universe in OpenMC it threw a bunch of errors saying "surface has no facets". Did you do some middle step?
- Do you have a preferred alternative scheme to the coarse mesh in the meantime? Just wondering if your experience suggests one scheme is more optimal than the others.

Thanks for your timely replies Patrick, you're a life saver!

**Reply by pshriwise** [reply]
No problem! It's high time we get all of these issues sorted out :) 

> By "with the existing mesh" do you mean you simply imported the exodus mesh, and then exported it to .h5m? I tried this once before but when I ran the geometry as my universe in OpenMC it threw a bunch of errors saying "surface has no facets". Did you do some middle step?

Yep! I imported the mesh as geometry and immediately exported a DAGMC file. I didn't attempt transport but the DAGMC ray fire test and point in volume test worked fine. This was with a different model that was tet-meshed, however. I see now yours is a hex mesh, which makes sense as to why you're getting the "no triangles" error.

> Do you have a preferred alternative scheme to the coarse mesh in the meantime? Just wondering if your experience suggests one scheme is more optimal than the others.

If it's possible to use a tet mesh for your exodus mesh, I'd probably recommend that for now. Otherwise, even without the coarse mesh settings, I think the tri mesher will be a good approximation of the original mesh. Some volume and surface area diagnostics would be warranted to ensure this is the case. I'm also working on a DAGMC toolbar that can be added to Cubit to streamline material assignment and boundary conditions. This toolbar will include a button for generating a report on the faceting tolerance of the current triangle mesh. I'll try to have it out soon and can send you an alpha version if you like for testing in the meantime.
